src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
+    {
+        /// <summary>
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(ProjectId projectId, string projectFullPath, string fileFullPath, CancellationToken cancellationToken);
```

Comment:
1 non source file returning 2 source files? since.. file is not actually meaningful, I guess they can return a file that contains multiple types in it?